Firestore/Source/Core/FSTFirestoreClient.m

```diff
@@ -249,6 +258,59 @@ - (void)removeListener:(FSTQueryListener *)listener {
   }];
 }
 
+- (void)getDocumentFromLocalCache:(FIRDocumentReference *)doc
+                       completion:(void (^)(FIRDocumentSnapshot *_Nullable document,
+                                            NSError *_Nullable error))completion {
+  [self.workerDispatchQueue dispatchAsync:^{
+    FSTMaybeDocument *maybeDoc = [self.localStore readDocument:doc.key];
+    if (maybeDoc) {
+      completion([FIRDocumentSnapshot snapshotWithFirestore:doc.firestore
+                                                documentKey:doc.key
+                                                   document:(FSTDocument *)maybeDoc
+                                                  fromCache:YES],
+                 nil);
+    } else {
+      completion(
+          nil, [NSError errorWithDomain:FIRFirestoreErrorDomain
+                                   code:FIRFirestoreErrorCodeUnavailable
+                               userInfo:@{
+                                 NSLocalizedDescriptionKey : @"Failed to get document from server.",
+                               }]);
+    }
+  }];
+}
+
+- (void)getDocumentsFromLocalCache:(FIRQuery *)query
+                        completion:(void (^)(FIRQuerySnapshot *_Nullable query,
+                                             NSError *_Nullable error))completion {
+  [self.workerDispatchQueue dispatchAsync:^{
+    FSTDocumentDictionary *docs = [self.localStore executeQuery:query.query];
+
+    __block FSTDocumentSet *documents =
+        [FSTDocumentSet documentSetWithComparator:query.query.comparator];
+    FSTDocumentSet *oldDocuments = documents;
+    [docs enumerateKeysAndObjectsUsingBlock:^(FSTDocumentKey *key, FSTDocument *value, BOOL *stop) {
+      documents = [documents documentSetByAddingDocument:value];
+    }];
+
+    FSTViewSnapshot *snapshot = [[FSTViewSnapshot alloc] initWithQuery:query.query
+                                                             documents:documents
+                                                          oldDocuments:oldDocuments
+                                                       documentChanges:@[]
+                                                             fromCache:YES
+                                                      hasPendingWrites:NO
```

Comment:
This seems like the wrong value. The local store will apply local mutations if present.

Also the documentChanges array being empty seems suspicious: the view processing code will normally make all the documents in the initial snapshot adds.

These are both examples of reasons why you shouldn't compose the view snapshots yourself.

Reply:
Done; ptal.